File: 2015/06.py
# --- Day 6: Probably a Fire Hazard ---

import re

with open("input/06.txt") as f:

  size = 1000
  lit = brightness = 0

  # create matrix
  grid = [[[0,0] for i in range(size)] for i in range(size)]

  for line in f:

    action = re.findall("(on|off|toggle)", line)[0]
    from_x,from_y,to_x,to_y = map(int, re.findall("([0-9]+)", line))

    for x in range (from_x, to_x + 1):
      for y in range (from_y, to_y + 1):
        if action == "on":
          grid[x][y][0] = 1                 # switch on
          grid[x][y][1] += 1                # increase brightness
        elif action == "off":
          grid[x][y][0] = 0                 # switch off
          if grid[x][y][1] > 0:
            grid[x][y][1] -= 1              # decrease brightness
        else:
          grid[x][y][0] = 1 - grid[x][y][0] # toggle
          grid[x][y][1] += 2                # increase brightness by 2

  # calculate brightness
  for x in range (0, size):
    for y in range (0, size):
      brightness += grid[x][y][1]
      if grid[x][y][0] == 1:
        lit += 1

  print lit
  print brightness

File: 2015/06.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

const gridSize = 1000

func format(text string) []int {
	split := strings.Split(text, ",")
	from, err := strconv.Atoi(split[0])
	if err != nil {
		log.Fatal(err)
	}

	to, err := strconv.Atoi(split[1])
	if err != nil {
		log.Fatal(err)
	}
	return []int{from, to}
}

func on(array [gridSize][gridSize]int, from []int, to []int, brightness bool) [gridSize][gridSize]int {
	for i := from[0]; i <= to[0]; i++ {
		for j := from[1]; j <= to[1]; j++ {
			if brightness {
				array[i][j]++
			} else {
				array[i][j] = 1
			}
		}
	}
	return array
}

func off(array [gridSize][gridSize]int, from []int, to []int, brightness bool) [gridSize][gridSize]int {
	for i := from[0]; i <= to[0]; i++ {
		for j := from[1]; j <= to[1]; j++ {
			if brightness && array[i][j] > 1 {
				array[i][j]--
			} else {
				array[i][j] = 0
			}
		}
	}
	return array
}

func toggle(array [gridSize][gridSize]int, from []int, to []int, brightness bool) [gridSize][gridSize]int {
	for i := from[0]; i <= to[0]; i++ {
		for j := from[1]; j <= to[1]; j++ {
			if brightness {
				array[i][j] += 2
			} else {
				value := array[i][j]
				if value == 0 {
					array[i][j] = 1
				} else {
					array[i][j] = 0
				}
			}
		}
	}
	return array
}

func main() {
	file, err := os.Open("input/06.txt")
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	var lights [gridSize][gridSize]int
	var brightLights [gridSize][gridSize]int

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		fields := strings.Fields(scanner.Text())
		action := fields[1]
		if action == "on" {
			lights = on(lights, format(fields[2]), format(fields[4]), false)
			brightLights = on(brightLights, format(fields[2]), format(fields[4]), true)
		} else if action == "off" {
			lights = off(lights, format(fields[2]), format(fields[4]), false)
			brightLights = off(brightLights, format(fields[2]), format(fields[4]), true)
		} else {
			lights = toggle(lights, format(fields[1]), format(fields[3]), false)
			brightLights = toggle(brightLights, format(fields[1]), format(fields[3]), true)
		}
	}

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	lit := 0
	brightness := 0

	for i := 0; i < gridSize; i++ {
		for j := 0; j < gridSize; j++ {
			if lights[i][j] == 1 {
				lit++
			}
			brightness += brightLights[i][j]
		}
	}

	fmt.Println(lit)
	fmt.Println(brightness)
}
